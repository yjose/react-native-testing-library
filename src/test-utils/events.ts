interface EventEntry {
  id: number;
  name: string;
  nativeEvent: any;
}

export function createEventToolkit() {
  let id = 0;
  const events: EventEntry[] = [];
  const handleEvent = (name: string) => {
    return (event: any) => {
      events.push({
        id,
        name,
        nativeEvent: event?.nativeEvent,
      });

      id += 1;
    };
  };

  return { events, handleEvent };
}
