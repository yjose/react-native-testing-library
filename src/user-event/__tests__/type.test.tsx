import * as React from 'react';
import { TextInput } from 'react-native';
import { createEventToolkit } from '../../test-utils/events';
import { render } from '../..';
import { userEvent } from '..';

function renderTextInputWithToolkit() {
  const { events, handleEvent } = createEventToolkit();

  const screen = render(
    <TextInput
      testID="input"
      onFocus={handleEvent('focus')}
      onBlur={handleEvent('blur')}
      onPressIn={handleEvent('pressIn')}
      onPressOut={handleEvent('pressOut')}
      onChange={handleEvent('change')}
      onChangeText={handleEvent('changeText')}
      onTextInput={handleEvent('textInput')}
      onKeyPress={handleEvent('keyPress')}
      onSelectionChange={handleEvent('selectionChange')}
      onSubmitEditing={handleEvent('onSubmitEditing')}
      onContentSizeChange={handleEvent('contentSizeChange')}
    />
  );

  return {
    ...screen,
    events,
  };
}

test('userEvent.type on TextInput', async () => {
  const { events, ...queries } = renderTextInputWithToolkit();

  await userEvent.type(queries.getByTestId('input'), 'ABC');

  const eventNames = events.map((e) => e.name);
  expect(eventNames).toEqual([
    'pressIn',
    'focus',
    'pressOut',
    'keyPress',
    'textInput',
    'change',
    'changeText',
    'selectionChange',
    'keyPress',
    'textInput',
    'change',
    'changeText',
    'selectionChange',
    'keyPress',
    'textInput',
    'change',
    'changeText',
    'selectionChange',
    'onSubmitEditing',
    'blur',
  ]);

  const entryEvents = events.slice(0, 3);
  const exitEvents = events.slice(events.length - 2);
  const typingEvents = events.slice(
    entryEvents.length,
    events.length - exitEvents.length
  );

  expect(entryEvents).toMatchSnapshot('entry events');
  expect(typingEvents).toMatchSnapshot('typing events');
  expect(exitEvents).toMatchSnapshot('exit events');
});

test('userEvent.type skips press events when `skipPress: true`', async () => {
  const { events, ...queries } = renderTextInputWithToolkit();
  await userEvent.type(queries.getByTestId('input'), 'abc', {
    skipPress: true,
  });

  const eventNames = events.map((e) => e.name);
  expect(eventNames).not.toContainEqual('pressIn');
  expect(eventNames).not.toContainEqual('pressOut');
  expect(events).toMatchSnapshot('events without press events');
});
