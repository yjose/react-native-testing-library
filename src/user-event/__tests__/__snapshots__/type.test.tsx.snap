// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userEvent.type on TextInput: entry events 1`] = `
[
  {
    "id": 0,
    "name": "pressIn",
    "nativeEvent": {
      "changedTouches": [],
      "identifier": 0,
      "locationX": 0,
      "locationY": 0,
      "pageX": 0,
      "pageY": 0,
      "target": 0,
      "timestamp": 1682759916812,
      "touches": [],
    },
  },
  {
    "id": 1,
    "name": "focus",
    "nativeEvent": {
      "target": 0,
    },
  },
  {
    "id": 2,
    "name": "pressOut",
    "nativeEvent": {
      "changedTouches": [],
      "identifier": 0,
      "locationX": 0,
      "locationY": 0,
      "pageX": 0,
      "pageY": 0,
      "target": 0,
      "timestamp": 1682759916814,
      "touches": [],
    },
  },
]
`;

exports[`userEvent.type on TextInput: exit events 1`] = `
[
  {
    "id": 18,
    "name": "onSubmitEditing",
    "nativeEvent": {
      "target": 0,
      "text": "ABC",
    },
  },
  {
    "id": 19,
    "name": "blur",
    "nativeEvent": {
      "target": 0,
    },
  },
]
`;

exports[`userEvent.type on TextInput: typing events 1`] = `
[
  {
    "id": 3,
    "name": "keyPress",
    "nativeEvent": {
      "key": "A",
    },
  },
  {
    "id": 4,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "",
      "range": {
        "end": 1,
        "start": 1,
      },
      "target": 0,
      "text": "A",
    },
  },
  {
    "id": 5,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "A",
    },
  },
  {
    "id": 6,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 7,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 1,
        "start": 1,
      },
    },
  },
  {
    "id": 8,
    "name": "keyPress",
    "nativeEvent": {
      "key": "B",
    },
  },
  {
    "id": 9,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "A",
      "range": {
        "end": 2,
        "start": 2,
      },
      "target": 0,
      "text": "AB",
    },
  },
  {
    "id": 10,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "AB",
    },
  },
  {
    "id": 11,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 12,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 2,
        "start": 2,
      },
    },
  },
  {
    "id": 13,
    "name": "keyPress",
    "nativeEvent": {
      "key": "C",
    },
  },
  {
    "id": 14,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "AB",
      "range": {
        "end": 3,
        "start": 3,
      },
      "target": 0,
      "text": "ABC",
    },
  },
  {
    "id": 15,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "ABC",
    },
  },
  {
    "id": 16,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 17,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 3,
        "start": 3,
      },
    },
  },
]
`;

exports[`userEvent.type skips press events when \`skipPress: true\`: events without press events 1`] = `
[
  {
    "id": 0,
    "name": "focus",
    "nativeEvent": {
      "target": 0,
    },
  },
  {
    "id": 1,
    "name": "keyPress",
    "nativeEvent": {
      "key": "a",
    },
  },
  {
    "id": 2,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "",
      "range": {
        "end": 1,
        "start": 1,
      },
      "target": 0,
      "text": "a",
    },
  },
  {
    "id": 3,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "a",
    },
  },
  {
    "id": 4,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 5,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 1,
        "start": 1,
      },
    },
  },
  {
    "id": 6,
    "name": "keyPress",
    "nativeEvent": {
      "key": "b",
    },
  },
  {
    "id": 7,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "a",
      "range": {
        "end": 2,
        "start": 2,
      },
      "target": 0,
      "text": "ab",
    },
  },
  {
    "id": 8,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "ab",
    },
  },
  {
    "id": 9,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 10,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 2,
        "start": 2,
      },
    },
  },
  {
    "id": 11,
    "name": "keyPress",
    "nativeEvent": {
      "key": "c",
    },
  },
  {
    "id": 12,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "ab",
      "range": {
        "end": 3,
        "start": 3,
      },
      "target": 0,
      "text": "abc",
    },
  },
  {
    "id": 13,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "abc",
    },
  },
  {
    "id": 14,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 15,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 3,
        "start": 3,
      },
    },
  },
  {
    "id": 16,
    "name": "onSubmitEditing",
    "nativeEvent": {
      "target": 0,
      "text": "abc",
    },
  },
  {
    "id": 17,
    "name": "blur",
    "nativeEvent": {
      "target": 0,
    },
  },
]
`;
